aggregator/election_mgr.go

```diff
@@ -0,0 +1,545 @@
+}
+
+func (mgr *electionManager) processGoalState(goalState goalState) {
+	currState := mgr.ElectionState()
+	newState := goalState.state
+	if currState == newState {
+		return
+	}
+	// NB(xichen): the follower to pending follower transition is conceptually invalid
+	// but because the leader election API currently does not verify states, if the client
+	// receives a follower state, it can only treat it as a "pendingFollower" state. This
+	// is no longer needed once the leader election logic is adapted to only notify clients
+	// of states that have been verified.
+	if currState == FollowerState && newState == PendingFollowerState {
+		return
+	}
+	mgr.electionStateWatchable.Update(newState)
+	mgr.metrics.stateChanges.Inc(1)
+	mgr.logger.Info(fmt.Sprintf("election state changed from %v to %v", currState, newState))
+}
+
+func (mgr *electionManager) verifyPendingFollower(watch xwatch.Watch) {
+	defer watch.Close()
+
+	for {
+		select {
+		case <-mgr.doneCh:
+			return
+		case <-watch.C():
+		}
+
+		currState := mgr.goalStateWatchable.Get().(goalState)
+		if currState.state != PendingFollowerState {
+			continue
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), mgr.changeVerifyTimeout)
+		continueFn := func(int) bool {
+			select {
+			case <-ctx.Done():
+				return false
+			default:
+				return true
+			}
+		}
+
+		var (
+			knownLeader bool
+			isLeader    bool
+		)
+		mgr.changeRetrier.AttemptWhile(continueFn, func() error {
+			leader, err := mgr.leaderService.Leader(mgr.electionKey)
+			if err != nil {
+				knownLeader = false
+				mgr.metrics.verifyLeaderErrors.Inc(1)
+				mgr.logError("error determining the leader", err)
+				return err
+			}
+
+			knownLeader = true
+			if leader == mgr.leaderValue {
+				isLeader = true
+				mgr.metrics.verifyLeaderNotChanged.Inc(1)
+				mgr.logError("leader has not changed", errLeaderNotChanged)
+				return errLeaderNotChanged
+			}
+			isLeader = false
+			return nil
+		})
+		cancel()
+
+		mgr.goalStateLock.Lock()
+		state := mgr.goalStateWatchable.Get().(goalState)
+		// If the latest goal state is different from the goal state before verification, this
+		// means the goal state has been updated since verification started and the pending follower
+		// to follower transition is no longer valid.
+		if !(state.id == currState.id && state.state == currState.state) {
+			mgr.metrics.verifyPendingChangeStale.Inc(1)
+			mgr.goalStateLock.Unlock()
+			continue
+		}
+		if !knownLeader {
+			mgr.metrics.verifyNoKnownLeader.Inc(1)
+			mgr.goalStateLock.Unlock()
+			continue
+		}
+		if isLeader {
+			mgr.setGoalStateWithLock(LeaderState)
+		} else {
+			mgr.setGoalStateWithLock(FollowerState)
+		}
+		mgr.goalStateLock.Unlock()
+	}
+}
+
+func (mgr *electionManager) resign(action resignAction) {
+	mgr.RLock()
+	if mgr.state != electionManagerOpen {
+		action.errCh <- errElectionManagerNotOpenOrClosed
+		mgr.RUnlock()
+		return
+	}
+	mgr.RUnlock()
+
+	if mgr.ElectionState() == FollowerState {
```

Comment:
Yes it's safe, `mgr.ElectionState()` does a `Get` on the underlying `electionStateWatchable`, which is thread-safe.